### PR TITLE
live migration support for SR-IOV, bug#868811

### PR DESCRIPTION
**Pull request by mmucha-redhat**
live migration support for SR-IOV, bug#868811

I confirm that this pull request was submitted according to the [contribution guidelines](/.github/CONTRIBUTING.md): (please @mention yourself to sign)

This pull request needs review by: (please @mention the reviewer if relevant)

initial commit of feature page for "Live Migration Support
For SRIOV" feature.

Signed-off-by: Martin Mucha <mmucha@redhat.com>

### CODE REVIEW

**Comment by mmucha-redhat** [comment]
@jhernand 
@dankenigsberg 
@mmirecki 
@myakove

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+@Type
+public interface VnicProfile extends Identified {
+    …
+    VnicPassThrough passThrough();
+    boolean migratable();
+```
+
+### GUI
+
+In UI, you need to flag passthrough nic as migratable to be able to do
```
Comment (dankenigsberg) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: in the example diaglog, it is not clear that "migratable" relates only to passthrough vNICs. regular vNICs are always migratable, and we should not confuse users to think otherwise. Can we "ungray" the migratable button only when passthrough is selected?

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+behavior user can specify, whether each passthrough nic is 'migratable' 
+or not. VM will can only be migrated, when all its nics passthrough 
+nics are marked as migratable. 
+
+Hot-plug can succeed only if there's available virtual function on 
+target host. To avoid possible race, migration will allocate virtual 
+function first, proceeding only if there was available one. 
+If there's none, VM won't be migrated.
+
+Also migration with SR-IOV is tricky and can fail. If that happens,
```
Comment (dankenigsberg) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: I'd rephrase this sentence.

Migration with SR-IOV vNICs is a tricky multi-state operation, and can fail. If the operation fails after the VM is already running on the destination, the VM would not be migrated back.

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    …
+    VnicPassThrough passThrough();
+    boolean migratable();
+```
+
+### GUI
+
+In UI, you need to flag passthrough nic as migratable to be able to do
+migration when SR-IOV is used.
+
```
Comment (dankenigsberg) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: Please add a section about guest-side support.

While migrating, the guest can notice that one of its vNICs has been unplugged, and communication is lost. To avoid that, VM admin should add two vNICs: an SR-IOV one for performance, and a virtio one for backup during migration. The guest should create a bonding (or teaming) device on top of these, so that user-space guest application can ignore the temporary disappearance of the SR-IOV device.

When migration finishes and the SR-IOV device is restored, it should be reattached to that bond. In the context of this feature, we would supply the el7 hooking mechanism to make it happen seamlessly.

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md (hunk 2 above)
Comment (mmucha-redhat) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: done.

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md (hunk 1 above)
Comment (mmucha-redhat) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: yes, only passthrough can be marked as migratable. 
screenshot updated.
Done.

**Comment by dankenigsberg** [comment]
I'm happy with the current state of feature page. @ydary, @mburman5 please approve!

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+migratable, or not marked as migratable. Other nics are always 
+migratable, thus migratable checkbox will be selected and greyed out
+when passthrough checkbox is not selected. 
+
+#### Setting migratable flag
+![Vnic profile with migratable flag png](vnicProfileWithMigratableFlag.png "Vnic profile with migratable flag png")
+
+###Guest-side support.
+   
+While migrating, the guest can notice that one of its vNICs has been
```
Comment (dankenigsberg) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: Here we ask our users to define an SRIOV network and an equivalent bridged network; to attach both to the VM; to allow mac spoofing; and to configure an in-guest bonding.

@mykaul wants us to put less burden on our users. Can you think of a grander approach, in which a user defines a "migratable passthrough network", and when attaching it to a VM, **two** vNICs are created, filter is set, and the guest is instructed to bond the two? It does not seem very easy; but maybe you can do it as a second phase of the feature as it is currently defined?

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
+---
+title: LiveMigrationSupportForSRIOV
+category: feature
+authors: mmucha
+wiki_category: Feature
+wiki_title: Features/LiveMigrationSupportForSRIOV
+wiki_revision_count: 12
+wiki_last_updated: 2015-10-14
```
Comment (mmirecki) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: Incorrect date

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+# Live Migration Support For SRIOV
+
+### Owner
+
+*   Name: [ Martin Mucha](User:mmucha)
+*   Email: mmucha@redhat.com
+
+## Summary
+Current support of SR-IOV in engine does not allow migration, which
```
Comment (mmirecki) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: Can you expand shortcuts on first usage?
SR-IOV (single root I/O virtualization)
VF (virtual function)
....

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+hot-plugging and hot-unplugging. We can employ this to enable currently 
+not supported migration. When VM is migrated to another host, its
+passthrough nic will be hot-unplugged, and related VF will be released.
+Then we can perform VM migration, and after that we'll perform hot-plug.
+Therefore, after migrating VM, there will be slight delay before nic 
+reappears in VM. 
+
+## Detailed Description
+
+Currently, VMs using SR-IOV nics cannot be migrate. To preserve this
```
Comment (mmirecki) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: migrate -> migrated

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+passthrough nic will be hot-unplugged, and related VF will be released.
+Then we can perform VM migration, and after that we'll perform hot-plug.
+Therefore, after migrating VM, there will be slight delay before nic 
+reappears in VM. 
+
+## Detailed Description
+
+Currently, VMs using SR-IOV nics cannot be migrate. To preserve this 
+behavior user can specify, whether each passthrough nic is 'migratable' 
+or not. VM will can only be migrated, when all its nics passthrough
```
Comment (mmirecki) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: VM will can ->  A VM will

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+reappears in VM. 
+
+## Detailed Description
+
+Currently, VMs using SR-IOV nics cannot be migrate. To preserve this 
+behavior user can specify, whether each passthrough nic is 'migratable' 
+or not. VM will can only be migrated, when all its nics passthrough 
+nics are marked as migratable. 
+
+Hot-plug can succeed only if there's available virtual function on
```
Comment (mmirecki) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: either use 'virtual function' or 'VF' everywhere

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+## Detailed Description
+
+Currently, VMs using SR-IOV nics cannot be migrate. To preserve this 
+behavior user can specify, whether each passthrough nic is 'migratable' 
+or not. VM will can only be migrated, when all its nics passthrough 
+nics are marked as migratable. 
+
+Hot-plug can succeed only if there's available virtual function on 
+target host. To avoid possible race, migration will allocate virtual
```
Comment (mmirecki) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: virtual -> a virtual

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+## Detailed Description
+
+Currently, VMs using SR-IOV nics cannot be migrate. To preserve this 
+behavior user can specify, whether each passthrough nic is 'migratable' 
+or not. VM will can only be migrated, when all its nics passthrough 
+nics are marked as migratable. 
+
+Hot-plug can succeed only if there's available virtual function on 
+target host. To avoid possible race, migration will allocate virtual 
+function first, proceeding only if there was available one.
```
Comment (mmirecki) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: was available one -> is one available

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md (hunk 5 above)
Comment (mmucha-redhat) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: done.

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md (hunk 7 above)
Comment (mmucha-redhat) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: done.

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md (hunk 6 above)
Comment (mmucha-redhat) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: done.

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md (hunk 8 above)
Comment (mmucha-redhat) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: Used: "A VM can only …"
Done.

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md (hunk 9 above)
Comment (mmucha-redhat) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: (not sure why that's important). Done.

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md (hunk 10 above)
Comment (mmucha-redhat) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: replaced with VF as requested. Done.

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md (hunk 11 above)
Comment (mmucha-redhat) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: a) not sure if past is incorrect, but I don't care
b) 'one' is here used as a placeholder (~ if there is available VF), not as a count. There can be more available VFs.
Done.

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md (hunk 1 above)
Comment (mburman5) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: If checking the 'migratable' checkbox it means that the 'passthorugh' should be checked as well, should be grayed out, correct?

Comment (mburman5) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: The screen shot attached isn't correct, a 'passthorugh' vNIC have no network filter, it should be grayed out and the no-filter-rule applied.

File: source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+#### Setting migratable flag
+![Vnic profile with migratable flag png](vnicProfileWithMigratableFlag.png "Vnic profile with migratable flag png")
+
+###Guest-side support.
+   
+While migrating, the guest can notice that one of its vNICs has been 
+unplugged, and communication is lost. To avoid that, VM admin should 
+add two vNICs: an SR-IOV one for performance, and a virtio one for 
+backup during migration. The guest should create a bonding (or teaming) 
+device on top of these, so that user-space guest application can ignore
```
Comment (mburman5) [inline source/develop/release-management/features/network/liveMigrationSupportForSRIOV.html.md]: As agreed with danken, we should provide a one good example of creating the bond in the guest.
